### ISSUE TITLE
No internet connection after after ~1 hour

### ISSUE THREAD

**Issue by bozegeduge** [julian-klode/dns66#105]
Phone:

Non rooted Sony Experia XA Ultra - Android 6.0

Problem:

After 1 hour or so, on WiFi (have not tested it on cellular data), I will have no internet access. I believe it may be a DNS issue, but ultimately I cannot access internet requests whether in Chrome or an application. 

Temporary resolution:

Turn off and on DNS66

Description:

Recently upgraded to the newest version of DNS66. I was having an issue with phone calls taking forever to dail/not dailing. Reseting DNS66 fixed the issue temporarily. Also this no internet access problem existed. Since upgrading I have not experienced the phone call issue, but while DNS66 is running after one hour or so, after a while i simply will not have internet access until I reset DNS66. There is no notification. All network settings are not modified except allowing only ipv4 on cellular data.

Thank you,

**Comment by ruku320** [comment]
I have the same issue on a 2nd generation Moto X (2014) with Android 6.0.

I would attempt to use my phone but it reacts as if there is no internet access.  After I turn DNS66 off and then back on (via long press in application) internet access works again.  I've only seen it happening on wifi.

This problem only occurs with the latest released version (0.4.1).  Don't see this when I revert back to 0.4.0 release.

**Comment by julian-klode** [comment]
That's strange. This should only happen when connectivity changes / when DNS66 reconnects. Are you sure there were no connectivity changes in that time period (switching between WiFis, WiFi to mobile data, reconnect WiFi, WiFi's connection reconnecting)?

0.4.1 had some refactoring, but that was mostly automatic, so I don't think that's causing any issues. It also changes the parser for host files which means reconnects are about 10 times faster, and that can cause issues because of Android bugs.

And I've been running 0.4.1 for 3 days in a row without any issues.

**Comment by serveravt** [comment]
the same issue, Moto X (1st generation 2013), 
in some time,  `AdVpnThread` stuck on the 255 line, 
`int result = FileHelper.poll(polls, -1);`

I don't know why, but in `deviceFd.events` there is no events :(

**Comment by julian-klode** [comment]
@serveravt Well, it should be stuck there, until a packet arrives.

The important question is when this happens. I've only ever seen this happen shortly after DNS66 reconnects. Android messes up if that goes too fast and no packets arrive in the device file descriptor anymore. A workaround for that is to make reconnects always take at least 2-3 seconds, so it has enough time to not get confused... If something else is going on, we'd be in bigger trouble.

**Comment by serveravt** [comment]
@julian-klode Yes, I understand that it waits for incoming packet on this line... 

But I am going to browser, trying to open any page, and it should trigger POLLIN event on deviceFileDescriptor, because browser should generate some DNS packets.... And in this time AdVpnThread still stuck on that 255th line :(  Restart VPN service of course helps...   Any suggestions on this? :)

**Comment by julian-klode** [comment]
@serveravt That's why I asked when this happens. That's the important part here. I have never seen this happen randomly, only after a reconnect (this also includes stuff like go to bed and it not working in the morning because the DSL reconnected during the night).

If you are browsing all the time, and DNS66 runs during that time without restarting, and it suddenly stops, that would be something completely different from the issue I know about, and that's what I'm trying to figure out.

**Comment by serveravt** [comment]
As I understand, it is happening randomly, it is not reconnection issue :(

**Comment by julian-klode** [comment]
@serveravt It would be *really* help to get a logcat as soon as you noticed it, possibly filtered for only those lines containing dns66.

**Comment by julian-klode** [comment]
Please try the experimental build from XDA: https://forum.xda-developers.com/showpost.php?p=71851497&postcount=442

It has some very experimental UI things from the ui-redesign branch (since replaced by ui-redesign2), but more importantly it adds a connection watchdog. Instead of waiting for packets infinitely, we now wait for 4s to about 1h  (exponents of base 4, increasing), then send a packet to the device, and then wait again, and if no packet has arrived within 7 seconds, we restart the connection.
